src/main/java/com/gojek/beast/worker/OffsetCommitWorker.java

```diff
@@ -13,72 +15,104 @@
-                    return new FailureStatus(new RuntimeException("Acknowledgement Timeout exceeded: " + offsetState.getAcknowledgeTimeoutMs()));
+            Instant startTime = Instant.now();
+            long start = clock.currentEpochMillis();
+            Map<TopicPartition, OffsetAndMetadata> partitionsCommitOffset = new HashMap<>();
+
+            int offsetClubbedBatches = 0;
+            while (true) {
+                Map<TopicPartition, OffsetAndMetadata> currentOffset = commitQueue.poll(queueConfig.getTimeout(), queueConfig.getTimeoutUnit());
+                if (stopped || clock.currentEpochMillis() - start > offsetState.getOffsetCommitTime()) {
+                    break;
+                }
+
+                if (currentOffset == null) {
+                    continue;
                 }
-                log.debug("waiting for {} acknowledgement for offset {}", defaultSleepMs, currentOffset);
-                sleep(defaultSleepMs);
-                statsClient.gauge("committer.queue.wait.ms", defaultSleepMs);
+
+                Instant commitQueuePollStartTime = Instant.now();
+                while (true) {
+                    if (offsetState.removeFromOffsetAck(currentOffset)) {
+                        commitQueue.remove(currentOffset);
+
+                        currentOffset.keySet().forEach(topicPartition -> {
+                            OffsetAndMetadata offsetAndMetadata = currentOffset.get(topicPartition);
+                            OffsetMetadata previousOffset = (OffsetMetadata) partitionsCommitOffset.getOrDefault(topicPartition, new OffsetMetadata(Integer.MIN_VALUE));
+                            OffsetMetadata newOffset = new OffsetMetadata(offsetAndMetadata.offset());
+                            if (previousOffset.compareTo(newOffset) < 0) {
+                                partitionsCommitOffset.put(topicPartition, newOffset);
+                            }
+                        });
+                        offsetClubbedBatches++;
+                        break;
+                    } else {
+                        if (offsetState.shouldCloseConsumer(partitionsCommitOffset)) {
+                            statsClient.increment("committer.ack.timeout");
+                            return new FailureStatus(new RuntimeException("Acknowledgement Timeout exceeded: " + offsetState.getAcknowledgeTimeoutMs()));
+                        }
+
+                        log.debug("waiting for {} acknowledgement for offset {}, {}, {}", defaultSleepMs, currentOffset, offsetState.partitionOffsetAckSize(), currentOffset);
+                        sleep(defaultSleepMs);
+                    }
+                }
+                statsClient.timeIt("committer.queue.wait.ms", commitQueuePollStartTime);
             }
-            statsClient.timeIt("committer.processing.time", start);
+            commit(partitionsCommitOffset);
+
+            statsClient.gauge("committer.clubbed.offsets", offsetClubbedBatches);
```

Comment:
What's the thought process behind using a gauge here?

Reply:
This metric will be published after OFFSET_COMMIT_TIME, and the value is the total number of batched offsets, this is not an indication of throughput, its a measure of how many batches were combined. Thus this has to be a gauge metric